Attempt to port KCL v2

trucnguyenlam opened the pull request:
This is my attempt to port the library to KCL v2. This PR seems to fix the problem of workers are requested to shutdown (which is mentioned in #19).

trucnguyenjgx committed:
Port to KCLv2

trucnguyenjgx committed:
Update dependency

trucnguyenjgx committed:
Adapt tests to aws sdk v2 and kcl v2

trucnguyenjgx committed:
Experiment with version

trucnguyenjgx committed:
Revert version experiment

trucnguyenlam committed:
Update sbt version and make test mock aws client

anikifoss commented:
@trucnguyenlam Thanks for the Kinesis 2.0 port! I had a quick look through the code, and it look good. My only note so far is that we have different coding styles, so we should add scalafmt with a default config to keep the style the same across the codebase.

Once the styling issues are resolved, I'll do a more thorough review during the next few weeks, and then we can merge this in.

trucnguyenlam commented:
@lexn82 thanks for your comment, could you please refer me the current project scalafmt config? I will try to set it up to align with the current project style.

trucnguyenjgx committed:
Fix test that use actor materializer as implicit

trucnguyenlam commented:
By the way, along with code style, I also need to update the README file (as it is for KCLv1).

trucnguyenlam commented on src/main/scala/com/contxt/kinesis/ConsumerConfig.scala:
```diff
@@ -0,0 +1,119 @@
+      }
+      .getOrElse(latestPos)
+
+    val kinesisClient = Option(kac).getOrElse(KinesisClientUtil.createKinesisAsyncClient(KinesisAsyncClient.builder()))
```
aws recommends to use [KinesisClientUtil](https://docs.aws.amazon.com/streams/latest/dev/kcl-migration.html) to create client

trucnguyenlam commented on src/it/scala/com/contxt/kinesis/TestStreamConfig.scala:
```diff
@@ -1,45 +1,61 @@
+      initialPositionInStream,
+      coordinatorConfig = Some(new CoordinatorConfig(applicationName).
+        shardConsumerDispatchPollIntervalMillis(idleTimeBetweenGetRecords.toMillis)),
+      retrievalConfig = Some(
```
create KCL client with disabled enhanced fanout by default

trucnguyenlam committed:
Apply scalafmt format IntelliJ style with scalafmt version 2.2.2

trucnguyenlam committed:
Update scalafmt version to latest

trucnguyenlam committed:
Modify README.md and little code tidy up.

trucnguyenlam commented:
@lexn82 I have applied a `scalafmt.conf` and changed README file, could you please have a look?

trucnguyenlam committed:
Change build to use scala version 2.12 by default

anikifoss commented:
@trucnguyenlam Thanks for the additional work on your patch and updating the readme! Let's bump the project version to `3.0.0` instead of just `2.1.0` since this is a pretty major change.

I've applied scalafmt with default config to master. If you run the scalafmt with the same settings, we should be in a good place to remove any style diffs. (I've noticed switching scalafmt styles leads to extra formatting, consider rebasing without intellij scalafmt changes, to make any merge conflicts go away.)

trucnguyenjgx committed:
Merge branch 'master' into port-kcl-v2

# Conflicts:
#	.scalafmt.conf
#	src/main/scala/com/contxt/kinesis/ConsumerStats.scala
#	src/main/scala/com/contxt/kinesis/KinesisAppId.scala
#	src/main/scala/com/contxt/kinesis/KinesisRecord.scala
#	src/main/scala/com/contxt/kinesis/KinesisSource.scala
#	src/main/scala/com/contxt/kinesis/MaterializerAsValue.scala
#	src/main/scala/com/contxt/kinesis/RecordProcessorImpl.scala
#	src/main/scala/com/contxt/kinesis/ShardCheckpointConfig.scala
#	src/main/scala/com/contxt/kinesis/ShardConsumerId.scala
#	src/test/scala/com/contxt/kinesis/KinesisRecordTest.scala
#	src/test/scala/com/contxt/kinesis/KinesisSourceFactoryTest.scala
#	src/test/scala/com/contxt/kinesis/MaterializerAsValueTest.scala
#	src/test/scala/com/contxt/kinesis/ShardCheckpointTrackerTest.scala

trucnguyenjgx committed:
Bump project version to 3.0.0 in README

trucnguyenlam commented:
@lexn82 Thanks for putting the default `scalafmt` in master branch, I have resolved the conflict and reapply `scalafmt`. Hopefully the code style will be consistent from now.
I also change version of the project in README to `3.0.0` as suggested (once it is released).

trucnguyenjgx committed:
Code style reformat due to unnecessary new line of old formatter

trucnguyenjgx committed:
Continue code style reformat due to unnecessary new line of old formatter continue

trucnguyenjgx committed:
Refactor leaseLost to use utility function shutdown

trucnguyenjgx committed:
Fix thread name not renderred

anikifoss commented on README.md:
````diff
@@ -129,6 +107,27 @@ calling `markProcessed()`, or the steam must be terminated with an exception. If
+## Consumer Configuration
+The Kinesis Consumer `ConsumerConfig` can be configured via HOCON configuration, which is common for Akka projects
+```hocon
+consumer {
````
`consumer` namespace at the top level may cause collision with other configurations (there may be other libraries doing consumer stuff and configured via HOCON). How about `kinesis-consumer` instead?

anikifoss commented on src/it/scala/com/contxt/kinesis/KinesisResourceManager.scala:
```diff
@@ -45,72 +66,112 @@ object KinesisResourceManager {
+    println("Waiting for " + tableName + " to become ACTIVE...")
+    val startTime = System.currentTimeMillis
+    val endTime = startTime + (10 * 60 * 1000)
+    while ({
```
Could be simplified to one line.

anikifoss commented on src/it/scala/com/contxt/kinesis/KinesisResourceManager.scala:
```diff
@@ -45,72 +66,112 @@ object KinesisResourceManager {
+  private def waitForTableToBecomeAvailable(tableName: String, dynamoDb: DynamoDbClient): Unit = {
+    println("Waiting for " + tableName + " to become ACTIVE...")
+    val startTime = System.currentTimeMillis
+    val endTime = startTime + (10 * 60 * 1000)
```
Is DynamoDb really that slow? 10 minute wait time seems excessive.

trucnguyenlam replied:
10minute is the maximum timeout, this loop can terminate early.

anikifoss commented on src/main/scala/com/contxt/kinesis/ConsumerConfig.scala:
```diff
@@ -0,0 +1,108 @@
+package com.contxt.kinesis
```
Stopping here for now.

trucnguyenjgx committed:
Update README in kinesis consumer configuration and simplify code in KinesisResourceManager

trucnguyenjgx committed:
Apply scalafmt

trucnguyenlam commented:
@lexn82: I've applied some small changes, please let me know if you are happy with those.

anikifoss commented:
@trucnguyenlam Thanks for the update!

I recently had a new addition to the family, so it's been challenging to find the time for coding. Thank you for your patience! Would it be alright with you if I could open a PR against your branch? Small changes are easier to fix in code rather than having the back and forth over the Github messages. That should speed things up a little bit.

trucnguyenlam commented:
@lexn82 Congratz on your good news! And surely, it's totally ok to open a PR to my branch.

bpg-dev commented:
Sorry, @lexn82, @trucnguyenlam, I did not notice this PR when was working in a parallel branch to bump scala version to 2.13. That branch have been merged and that caused conflicts here.
I'll fix them and make a PR to trucnguyenlam:port-kcl-v2

bpg-dev commented:
> I'll fix them and make a PR to trucnguyenlam:port-kcl-v2

Here you go: https://github.com/trucnguyenlam/kcl-akka-stream/pull/1

trucnguyenlam commented:
@bpg thanks for making a PR to my branch, I'll have a look into that and get it merged ASAP

bpg-dev committed:
Fix merge conflicts with `StreetContxt/kcl-akka-stream` master (#1)

* Bump up SBT & Scala versions (#23)

- SBT 1.3.9, Scala 2.13.1
- Cross-compile for 2.12, 2.13, remove 2.11 support
- Fix compatibility issues / warnings

* add a few changes missed during conflict resolution

trucnguyenjgx committed:
Resolve merge conflict with upstream master

trucnguyenlam commented:
@bpg I've merged your PR and resolve the conflict. Please have a look.

trucnguyenjgx committed:
Use MapView instead of map in Inspectable

anikifoss commented:
@bpg @agenovese I could not help with this project for the last couple of month. Could you take over and review the PR?

agenovese commented:
I'll see if I can put together the time, as you can imagine things have been really hectic lately.

My Apologies @trucnguyenlam for the delay, and I want to express my appreciation for the PR. I will do my best to get to this in the next week or two.

trucnguyenlam commented:
@agenovese Thanks, let's get this done in two weeks 👍

elise-scx reviewed (approved):
Looks good to me! For fun I updated one of my projects to use this version, and it works nicely. Updating does require some extra code changes, so incrementing the major version to 3 was a good call.

angelo-streetcontxt reviewed (approved):
Looks good to me. 

Merging now and will tag and publish a release later today.

I want to say thank you again Truc, and my apologies for how long this took to get merged.

trucnguyenlam commented:
Thanks a lot for getting this done 👍

trucnguyenlam commented:
@angelo-streetcontxt by the way, do you know how long this would take (roughly) for the artifact to appear on maven repo [here](https://mvnrepository.com/artifact/com.streetcontxt/kcl-akka-stream)? I haven't uploaded any artifact to maven yet so I'm not clear on that :)

angelo-streetcontxt commented:
To be honest, I didn't even realize we had it showing up there. As far as I knew, assuming you have our bintray repo setup, the artefact should have become available pretty much immediately.

In our builds we have:

`resolvers in ThisBuild += Resolver.bintrayRepo("streetcontxt", "maven")`

and 

`libraryDependencies += "com.streetcontxt" %% "kcl-akka-stream" % "3.0.0"`

I'll see if I can figure out how/where those additional releases got to a more "official" server

angelo-streetcontxt commented:
Ok, after a bit of digging, that site must be indexing out repo (the one I mentioned above). 
Not sure how frequently they crawl the repositories, and it doesn't seem to be stated on their site as far as I can tell.

If you're using SBT to download the dependency, the code above should work, if you are downloading it manually you should be able to get the jar here:

https://dl.bintray.com/streetcontxt/maven/com/streetcontxt/kcl-akka-stream_2.11/3.0.0/
https://dl.bintray.com/streetcontxt/maven/com/streetcontxt/kcl-akka-stream_2.12/3.0.0/
https://dl.bintray.com/streetcontxt/maven/com/streetcontxt/kcl-akka-stream_2.13/3.0.0/

trucnguyenlam commented:
Just tested on your repo
```
resolvers in ThisBuild += Resolver.bintrayRepo("streetcontxt", "maven")
```

works perfectly, thanks 👍
